src/common/stats/MetricsSerializer.cpp

```diff
@@ -0,0 +1,170 @@
+/* Copyright (c) 2019 vesoft inc. All rights reserved.
+ *
+ * This source code is licensed under Apache 2.0 License,
+ * attached with Common Clause Condition 1.0, found in the LICENSES directory.
+ */
+
+#include "MetricsSerializer.h"
+
+namespace nebula {
+namespace stats {
+
+void PrometheusSerializer::annotate(
+        std::ostream& out, const std::string& metricName, const std::string& metricType) const {
+    out << "# HELP " << metricName << " Record all " << metricType << " about nebula" << "\n";
+    out << "# TYPE " << metricName << " " << metricType << "\n";
+}
+
+// Write a double as a string, with proper formatting for infinity and NaN
+void PrometheusSerializer::writeValue(std::ostream& out, double value) const {
+    if (std::isnan(value)) {
+        out << "Nan";
+    } else if (std::isinf(value)) {
+        out << (value < 0 ? "-Inf" : "+Inf");
+    } else {
+        auto saved_flags = out.setf(std::ios::fixed, std::ios::floatfield);
+        out << value;
+        out.setf(saved_flags, std::ios::floatfield);
+    }
+}
+
+void PrometheusSerializer::writeValue(std::ostream& out, const int64_t value) const {
+    if (std::isnan(value)) {
+        out << "Nan";
+    } else if (std::isinf(value)) {
+        out << (value < 0 ? "-Inf" : "+Inf");
+    } else {
+        out << value;
+    }
+}
+
+void PrometheusSerializer::writeValue(std::ostream& out, const std::string& value) const {
+    for (auto c : value) {
+        switch (c) {
+        case '\n':
+            out << '\\' << 'n';
+            break;
+
+        case '\\':
+            out << '\\' << c;
+            break;
+
+        case '"':
+            out << '\\' << c;
+            break;
+
+        default:
+            out << c;
+            break;
+        }
+    }
+}
+
+template<typename T>
+void PrometheusSerializer::writeHead(std::ostream& out, const std::string& family,
+        const std::vector<Label>& labels, const std::string& suffix,
+        const std::string& extraLabelName,
+        const T extraLabelValue) const {
+    out << family << suffix;
+    if (!labels.empty() || !extraLabelName.empty()) {
+        out << "{";
+        const char* prefix = "";
+
+        for (auto& lp : labels) {
+            out << prefix << lp.name << "=\"";
+            writeValue(out, lp.value);
+            out << "\"";
+            prefix = ",";
+        }
+        if (!extraLabelName.empty()) {
+            out << prefix << extraLabelName << "=\"";
+            writeValue(out, extraLabelValue);
+            out << "\"";
+        }
+        out << "}";
+    }
+    out << " ";
+}
+
+// Write a line trailer: timestamp
+void PrometheusSerializer::writeTail(std::ostream& out, const std::int64_t timestampMs) const {
+    if (timestampMs != 0) {
+        out << " " << timestampMs;
+    }
+    out << "\n";
+}
+
+
+void PrometheusSerializer::prometheusSerialize(std::ostream& out, StatsManager& sm) const {
```

Comment:
Use `const` to qualify the function parameter type `sm`?

Reply:
The mutex require non-const.